### PR TITLE
feat: support OCI image manifest

### PR DESCRIPTION
**Pull request by patrickzheng200**
This PR adds OCI image manifest support to notation CLI according to this [spec](https://github.com/notaryproject/notation/blob/main/specs/commandline/sign.md).

If using OCI artifact manifest to store signatures, notation requires the registry to support Referrers API as well.

Tested with [COSE, JWS] + [OCI artifact manifest, OCI image manifest] on ACR.

This PR would also resolve [issue](https://github.com/notaryproject/notaryproject/issues/219).

### CODE REVIEW

File: cmd/notation/registry.go
Comment (JeyJeyGao) [inline cmd/notation/registry.go:44]: Can we add a comment to mention that the `ociImageManifest` only influences on signing logic but not the verifying and listing logic.

File: cmd/notation/sign.go
Comment (JeyJeyGao) [inline cmd/notation/sign.go:88]: Can move the section out of runSign() to make the function clear. What about move it to

```go
		RunE: func(cmd *cobra.Command, args []string) error {
            // Here
			return runSign(cmd, opts)
		},
```

File: cmd/notation/sign.go
Comment (JeyJeyGao) [inline cmd/notation/sign.go:21]: do we need the empty line?

File: cmd/notation/registry.go
Comment (patrickzheng200) [reply cmd/notation/registry.go:44]: It's an unexported helper function though. Will add a description comment for this argument.

File: cmd/notation/registry.go
Comment (JeyJeyGao) [inline cmd/notation/registry.go:94]: Why not let oras-go automatically determine whether use referrers api or not?

Comment (patrickzheng200) [reply cmd/notation/registry.go:94]: Because oras-go does not have this capability yet. Notation needs to explicitly ping Referrers API.

File: specs/commandline/sign.md
Comment (shizhMSFT) [inline specs/commandline/sign.md:35]: According to the spec, should it be `--image-spec`?

File: internal/slices/slices.go
Comment (shizhMSFT) [inline internal/slices/slices.go:30]: nit: The function name is acceptable but still weird. It should named as `Contains` but the name is occupied. How about `ContainerElement`?

File: cmd/notation/registry.go
Comment (shizhMSFT) [inline cmd/notation/registry.go:87]: This block of code seems duplicated. Can we refactor a bit?

File: cmd/notation/sign.go
Comment (shizhMSFT) [inline cmd/notation/sign.go:21]: nit:
```suggestion
	imageSpecArtifact = "v1.1-artifact"
	imageSpecImage    = "v1.1-image"
```

File: cmd/notation/sign.go
Comment (shizhMSFT) [inline cmd/notation/sign.go:98]: nit: code can be simplified
```suggestion
	ociImageManifest := cmdOpts.imageSpec == imageManifest
```

File: cmd/notation/sign.go
Comment (priteshbandi) [inline cmd/notation/sign.go:72]:
```suggestion
				return fmt.Errorf("image spec must be one of the following %v but got %s", supportedImageSpec, opts.imageSpec)
```

File: cmd/notation/sign_test.go
Comment (priteshbandi) [inline cmd/notation/sign_test.go:26]: We should be consistent on using `artifact-type` or `imageSpec` everywhere.

**Review by priteshbandi — COMMENTED** [review]
This PR contains changes only for [sign](https://github.com/notaryproject/notaryproject/issues/219) command, will there be a separate PR for [verify](https://github.com/notaryproject/notaryproject/issues/220) command? 
Or is that not required because oras-go already handles fallback for verify ?

File: cmd/notation/registry.go
Comment (JeyJeyGao) [reply cmd/notation/registry.go:94]: https://github.com/oras-project/oras-go/blob/main/registry/remote/repository.go#L532
I think oras-go has the capability, right?

File: specs/commandline/sign.md
Comment (yizha1) [reply specs/commandline/sign.md:35]: @priteshbandi I recall that you did agree with the name `--image-spec` during spec review, see https://github.com/notaryproject/notation/pull/502#discussion_r1069095292.
It is more flexible and extensible by this design `--image-spec <spec version>-<manifest type>`  that users can specify the spec version and the manifest type that they want to use. There could be new types in the future. On the other hand, the name of `--artifact-type` could cause confusion, since there is a field named `artifactType` defined in `OCI artifact manifest` spec. User may think this flag `--artifact-type` is to be used to configure the filed `artifactType`.
I suggest we keep the current definition in the spec.

File: internal/slices/slices.go
Comment (JeyJeyGao) [inline internal/slices/slices.go:30]: Can we add a unit test for it, so it will not reduce the code coverage.

File: cmd/notation/registry.go
Comment (Wwwsylvia) [inline cmd/notation/registry.go:93]: Could you help me to understand the scenarios here? Should we list the valid scenarios in the comment? 
From the code it looks like:
1. If OCI artifact manifest is used
  - Referrers API must be available
  - Must not fallback to referrers tag schema (why?)

2. If OCI Image manifest is used
  - Referrers API and tag schema are both OK (?)

Comment (patrickzheng200) [reply cmd/notation/registry.go:93]: This change is based on this community [discussion](https://cloud-native.slack.com/archives/CQUH8U287/p1674050115958189?thread_ts=1673362612.471549&cid=CQUH8U287) that notation will not implement hybrid support, for instance, a registry supports OCI artifact manifest but not the Referrers API. 
Therefore, when user wants to store signatures in OCI artifact manifest, we require the Referrers API as well (and this is the default behavior of notation). 
On the other hand, I haven't seen a registry that supports Referrers API but only OCI image manifest. Hence, when user chooses OCI image manifest, notation could utilize oras's fallback logic and use the tag schema. But to be safer, yes, I think setting `remoteRepo.SetReferrersCapability(false)` on OCI image manifest is a better way here. I will update the code to reflect this.

File: cmd/notation/registry.go
Comment (patrickzheng200) [reply cmd/notation/registry.go:94]: Currently, oras-go does not have this capability yet. See https://github.com/oras-project/oras-go/issues/423 for more details.

File: cmd/notation/registry.go
Comment (Wwwsylvia) [reply cmd/notation/registry.go:93]: Thanks @patrickzheng200  for the explanation.

> But to be safer, yes, I think setting remoteRepo.SetReferrersCapability(false) on OCI image manifest is a better way here.

To my understanding, it seems that it won't hurt to use the Referrers API for OCI image manifests. But if the Referrers API is known to be unavailable, setting `remoteRepo.SetReferrersCapability(false)` will bring better performance since the call to Referrers API will be saved.

File: specs/commandline/sign.md
Comment (patrickzheng200) [reply specs/commandline/sign.md]: I changed back to `--image-spec` for now based on comments above.

**Comment by patrickzheng200** [comment]
> This PR contains changes only for [sign](https://github.com/notaryproject/notaryproject/issues/219) command, will there be a separate PR for [verify](https://github.com/notaryproject/notaryproject/issues/220) command? Or is that not required because oras-go already handles fallback for verify ?

Yes, this change only take effect on the Sign process. For Verification, we should already have the correct logic based on community discussion.

File: cmd/notation/registry.go
Comment (patrickzheng200) [reply cmd/notation/registry.go]: > Thanks @patrickzheng200 for the explanation.
> 
> > But to be safer, yes, I think setting remoteRepo.SetReferrersCapability(false) on OCI image manifest is a better way here.
> 
> To my understanding, it seems that it won't hurt to use the Referrers API for OCI image manifests. But if the Referrers API is known to be unavailable, setting `remoteRepo.SetReferrersCapability(false)` will bring better performance since the call to Referrers API will be saved.

Just had another discussion with Shiwei, Notation would still enforce two paths during Signing:
1. OCI artifact manifest uses the Referrers API
2. OCI image manifest uses the Referrers Tag Schema
(So basically, we do not support hybrid case on both ways.)

File: cmd/notation/verify.go
Comment (JeyJeyGao) [inline cmd/notation/verify.go:59]: why the shorthand "c" was removed?

Comment (patrickzheng200) [reply cmd/notation/verify.go:59]: This was a suggestion by @priteshbandi in this [comment](https://github.com/notaryproject/notation/pull/509#discussion_r1081951504).

Comment (JeyJeyGao) [reply cmd/notation/verify.go:59]: If we have multiple plugin config key=value pairs, a shorthand is helpful. 
```
notation sign xxxxx \
-c key1=value1
-c key2=value2
```
@priteshbandi is this a rational use case that we need the shorthand for plugin-config flag?

File: cmd/notation/registry.go
Comment (shizhMSFT) [inline cmd/notation/registry.go:94]: nit: code
```go
	remoteRepo := &remote.Repository{
		Client:    authClient,
		Reference: ref,
		PlainHTTP: plainHTTP,
	}
	return remoteRepo, nil
```
can be simplified as
```go
	return &remote.Repository{
		Client:    authClient,
		Reference: ref,
		PlainHTTP: plainHTTP,
	}, nil
```

**Review by shizhMSFT — APPROVED** [review]
LGTM

File: specs/commandline/sign.md
Comment (jlbutler) [reply specs/commandline/sign.md]: FWIW, I agree with @priteshbandi on this. I think users specifying both an image spec version and artifact type feels like too much implementation detail exposure.

Comment (patrickzheng200) [reply specs/commandline/sign.md]: @priteshbandi @jlbutler Another option here is making it a boolean flag, for example, `--legacy-artifact-type`. Omitting the flag means using OCI artifact manifest to store signatures. Setting the flag means using OCI image manifest. What do you think?

File: cmd/notation/registry.go
Comment (Wwwsylvia) [reply cmd/notation/registry.go]: Got it!

File: cmd/notation/verify.go
Comment (JeyJeyGao) [reply cmd/notation/verify.go:59]: Update: I think "--plugin-config" is fine.

File: specs/commandline/sign.md
Comment (FeynmanZhou) [reply specs/commandline/sign.md]: I vote for `--image-spec <spec version>-<manifest type>`. By specifying the OCI image spec version, it will be much clear as the artifact/image manifest may differentiate by spec versions.

Just FYI, ORAS has a similar situation with specifying a tag to custom the pushing behavior of the manifest type, see https://github.com/oras-project/oras/blob/main/docs/proposals/compatibility-mode.md#force-uploading-a-specific-manifest-type-using-a-flag---image-spec

**Review by priteshbandi — APPROVED** [review]
LGTM with pending flag name [discussion](https://cloud-native.slack.com/archives/CQUH8U287/p1675391977693439). For merge, I would suggest wait for that discussion to conclude and then based on it update the flag name.

File: specs/commandline/sign.md
Comment (toddysm) [reply specs/commandline/sign.md]: I think both `--image-spec` and `--artifact-type` as flag are confusing. I am struggling to understand the meaning of both. Also, from the help text above the expectation seems to be that this parameter will expect some value like `v1.0` or `v1.1`. This is too much details for somebody who doesn't even know what is the difference between OCI 1.0 and OCI 1.1 or has no knowledge of how different registries store the signature. 

The flow is the following:
- User tries to store the signature into the registry and gets an error.
- Based on the descriptiveness of the error, the user may want to retry the operation with the switch

There are two parts here:
- First, we need to make sure that the error the user sees is descriptive enough so they know what is happening. I would expect to get some error like: **Saving the signature failed. It is possible the registry doesn't support the format the signature is stored. Try --THE_SWITCH to use the legacy format for storing the signature**
- The name of the switch should be more user friendly and not requiring the user to know anything about the OCI spec. What comes to my mind is something like `--use-legacy-format`

Comment (yizha1) [reply specs/commandline/sign.md]: @toddysm Thanks. It makes sense. There is an existing flag named `--signature-format` for `JWS` or `COSE` envelope. So, it's better not to use the word `format` to avoid confusion. How about `--use-legacy-manifest`. Here is the summary for current proposals:
1. flag name `--signature-manifest`, values could be `1.0` or `1.1`
2. flag name `--signature-manifest`, values could be `image` or `artifact`
3. flag name `--use-legacy-format`, this is a Boolean type of flag
4. flag name `--use-legacy-manifest`, this is a Boolean type of flag

Comment (yizha1) [reply specs/commandline/sign.md]: One more thing, is it appropriate to use the wording `legacy` before OCI spec cut a 1.1 release?

Comment (toddysm) [reply specs/commandline/sign.md]: I see. So, what about `--use-image-manifest` Boolean flag? This is quite explicit, and it doesn't require deep knowledge of the OCI spec versioning.

Comment (lachie83) [reply specs/commandline/sign.md]: General agreement on the Notary community call on Feb 6 was to proceed with option 2 above using string values of `image` and `artifact`, `artifact` will be the default. cc @patrickzheng200

**Comment by patrickzheng200** [comment]
The current build would fail due to dependency change in notation-go: https://github.com/notaryproject/notation-go/pull/271

Please review https://github.com/notaryproject/notation-go/pull/271 first.

File: cmd/notation/registry.go
Comment (JeyJeyGao) [inline cmd/notation/registry.go:68]: Looks like it was not used?

**Comment by patrickzheng200** [comment]
Dependency updated. PR ready to be reviewed/merged.

File: cmd/notation/registry.go
Comment (patrickzheng200) [reply cmd/notation/registry.go]: Removed.

File: cmd/notation/sign.go
Comment (priteshbandi) [inline cmd/notation/sign.go:67]: IMO we dont need to show examples for all the signature envelope

```suggestion
Example - Sign an OCI artifact and use OCI image manifest to store the signature:
  notation sign --signature-manifest image <registry>/<repository>@<digest>
```

File: cmd/notation/registry.go
Comment (shizhMSFT) [inline cmd/notation/registry.go:204]: The error message can be simplified as
```suggestion
		errMsg := "Target registry does not support the Referrers API. Try the flag `--signature-manifest image` for backwards compatibility", err)
```
where the detailed error message can be in the log.

File: cmd/notation/sign.go
Comment (shizhMSFT) [inline cmd/notation/sign.go:117]: nit: The error message can be simplified.
```suggestion
			return fmt.Errorf("%v. Target registry seems not support OCI artifact manifest. Try the flag `--signature-manifest image` to store signatures using OCI image manifest", err)
```

Comment (priteshbandi) [reply cmd/notation/sign.go:117]: +1,

`Target registry does not seem to support OCI artifact manifest. Try the flag `--signature-manifest image` to store signatures using OCI image manifest`

File: cmd/notation/registry.go
Comment (priteshbandi) [reply cmd/notation/registry.go:204]: To be consistent should we use `Try the flag --signature-manifest image to store signatures using OCI image manifest` here ?

File: cmd/notation/registry.go
Comment (shizhMSFT) [reply cmd/notation/registry.go:200]: We probably need to have a comment block to describe this path.

File: cmd/notation/registry.go
Comment (patrickzheng200) [reply cmd/notation/registry.go:204]: I think we could combine both, i.e. `Try the flag --signature-manifest image to store signatures using OCI image manifest for backwards compatibility`. The benefit to include `for backwards compatibility` is that a user might not know what `OCI image manifest` is. Adding `for backwards compatibility` shows the purpose behind this flag.

**Review by shizhMSFT — APPROVED** [review]
LGTM

**Review by priteshbandi — APPROVED** [review]
LGTM, you will need to resolve conflict before change can be merged

**Pull request merged**
